cupy/_core/include/cupy/numpystring.h

```diff
@@ -0,0 +1,150 @@
+    // TODO: 21 is large enough for 64bit integers but this should use
+    //       std::numeric_limits or similar?
+    char digits[21];
+
+    bool const is_negative = value < 0;
+    IntT absval = is_negative ? -value : value;
+
+    int digits_idx = 0;
+    do {
+        digits[digits_idx++] = '0' + absval % (IntT)10;
+        // next digit
+        absval = absval / (IntT)10;
+    } while (absval != 0);
+
+    CharT *ptr = ptr_orig;
+    if (is_negative) {
+        *ptr++ = '-';
+        strlen--;
+    }
+    // digits are backwards, reverse the string into the output
+    while (digits_idx-- > 0 && strlen-- > 0) {
+        *ptr++ = digits[digits_idx];
+    }
+
+    /* zero fill unused chunk */
+    while (strenlen--) {
+        *ptr++ = 0;
+    }
+}
+
+
+/* The code below is heavily inspired by the cuDF version */
+template<typename IntT, typename CharT>
+__host__ __device__ inline IntT
+string_to_integer(int strlen, CharT *ptr)
+{
+    IntT value = 0;
+
+    if (strlen == 0) {
+        return value;
+    }
+    int sign = 1;
+    if (*ptr == '-' || *ptr == '+') {
+        sign = (*ptr == '-' ? -1 : 1);
+        ++ptr;
+        --strlen;
+    }
+    for (int idx = 0; idx < strlen; idx++, ptr++) {
+        CharT chr = *ptr;
+        if (chr < '0' || chr > '9') {
+            // TODO: Maybe it would be good to set INT_MIN/UINT_MAX for real
+            //       invalid numbers (at this point could be NULL terminated).
+            break;
+        }
+        value = (value * (IntT)10) + static_cast<IntT>(chr - '0');
+    }
+    return value * sign;
+}
+
+
+template<typename T, int maxlen_>
+class NumPyString {
+public:
+    static const int maxlen = maxlen_;
+    // TODO: 0 is possible, but C doesn't like it here (for good reasons)
+    //       there may be a way to tell C++ that empty is OK/possible?
+    T data[maxlen_ ? maxlen_ : 1];
+
+    __host__ __device__ int strlen() {
+        int len = maxlen;
+        while (len > 0 && this->data[len-1] == 0) {
+            len--;
+        }
+        return len;
+    }
+
+    __host__ __device__ NumPyString () {}
+
+    template<typename OT, int Olen>
+    __host__ __device__ NumPyString (const NumPyString<OT, Olen> &other)
+    {
+        // TODO: This is very unsafe (as it just casts)
+        for (int i = 0; i < this->maxlen; i++) {
+            this->data[i] = other[i];
+        }
+    }
+
+    __host__ __device__ T operator[](int i) const {
+        /* Allowing too large `i` for easier handling of different length */
+        if (i < this->maxlen) {
+            return this->data[i];
+        }
+        return 0;
+    }
+
+    template<typename OT, int Olen>
+    __host__ __device__ NumPyString& operator=(const NumPyString<OT, Olen> &other)
+    {
+        // NOTE: Unlike NumPy, we just cast U->S (unsafe).
+        for (int i = 0; i < this->maxlen; i++) {
+            this->data[i] = other[i];
+        }
+        return *this;
+    }
+
+    // TODO: Howto best template it for all integers?!
```

Comment:
One way I can think of is to make `long` here as a template parameter T, and use std::enable_if to enforce that only integer T is allowed.

Reply:
Yeah, I need to strengthen my C++ foo in this area.  I suspect the problem I had with `enable_if` was/may have been that we probably need CCCL to have the type traits header available?

Reply:
Ah, we have it here
https://github.com/cupy/cupy/blob/c2133b1b5d723893b06e0ae4feec1bfd2098618d/cupy/_core/include/cupy/carray.cuh#L10-L26
so that we don't need CCCL (for the time being). I don't recall if jitify is needed or not.